### setup MLflow

In [1]:
!pip install mlflow pyngrok lightgbm xgboost catboost categories -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 1.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.2/50.2 kB 2.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.5/43.5 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 59.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 64.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 27.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.8/147.8 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.0/212.0 kB 10.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.3/86.3 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.2/132.2 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 907.5/

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import mlflow
import mlflow.sklearn
import os

In [3]:
# Mengaktifkan autologging untuk library scikit-learn agar metrik tercatat otomatis
mlflow.sklearn.autolog()

In [4]:
# menentukan nama eksperimen di MLflow
experiment_name = "Bank_Marketing_Senior_Project"
if not mlflow.get_experiment_by_name(experiment_name):
    mlflow.create_experiment(experiment_name)
mlflow.set_experiment(experiment_name)

print("Setup MLflow Berhasil!")

2026/06/04 01:22:44 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/06/04 01:22:44 INFO mlflow.store.db.utils: Updating database tables


Setup MLflow Berhasil!


### Load data

In [5]:
# 1. Mount Drive (Hanya perlu sekali per sesi)
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
import pandas as pd

path = '/content/drive/My Drive/CSV/Bank Marketing.csv'
df = pd.read_csv(path, sep=';')

print(f"Bentuk Dataset: {df.shape}")
df.head()

Bentuk Dataset: (41188, 21)


,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


### EDA
Melihat ketimpangan kelas target (y) dan Distribusi umur nasabah

In [7]:
# Run di MLflow untuk mencatat aktivitas EDA
with mlflow.start_run(run_name="Exploratory_Data_Analysis"):

    # 1. Analisis Ketimpangan Kelas (Target y)
    plt.figure(figsize=(6,4))
    sns.countplot(x='y', data=df, palette='Set2')
    plt.title('Distribusi Target (Subscription Deposit)')
    plt.xlabel('Apakah Langganan? (y)')
    plt.ylabel('Jumlah Nasabah')

    # Simpan plot secara lokal
    target_plot_path = "target_distribution.png"
    plt.savefig(target_plot_path)
    plt.close()

    # 2. Analisis Distribusi Umur berdasarkan Target
    plt.figure(figsize=(10,5))
    sns.histplot(data=df, x='age', hue='y', multiple='stack', kde=True, palette='coolwarm')
    plt.title('Distribusi Umur Nasabah vs Keputusan Deposito')

    age_plot_path = "age_distribution.png"
    plt.savefig(age_plot_path)
    plt.close()

    # --- LOGGING KE MLFLOW ---
    # mencatat informasi dasar berupa parameter deskriptif data
    mlflow.log_param("total_rows", df.shape[0])
    mlflow.log_param("total_columns", df.shape[1])
    mlflow.log_param("imbalance_ratio", round(df['y'].value_counts(normalize=True)['no'], 4))

    # mengunggah file gambar tadi sebagai Artifact ke dalam MLflow
    mlflow.log_artifact(target_plot_path)
    mlflow.log_artifact(age_plot_path)

    # Bersihkan file lokal setelah diunggah ke MLflow
    os.remove(target_plot_path)
    os.remove(age_plot_path)

    print("EDA selesai dan visualisasinya sudah aman dicatat di MLflow!")

/tmp/ipykernel_987/3510685052.py:6: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.countplot(x='y', data=df, palette='Set2')


EDA selesai dan visualisasinya sudah aman dicatat di MLflow!


### melihat dashboard

In [8]:
from pyngrok import ngrok
import os
import time

# 1. Pastikan matikan semua sisa proses ngrok lama agar port bersih
ngrok.kill()
os.system("pkill -f mlflow") # Mematikan server mlflow lama yang macet

time.sleep(2) # Tunggu 2 detik agar port benar-benar lepas

# 2. memasukan token Ngrok
NGROK_TOKEN = "3Ec7HLSq4ZalTatQiTRbZmDu0hH_3fZabaRZffX5TZAnKkoz9"
ngrok.set_auth_token(NGROK_TOKEN)

# 3. Jalankan server MLflow dengan opsi bypass host checking lewat gunicorn
# menambahkan --expose-config agar mlflow lebih fleksibel di lingkungan proxy
os.system("mlflow ui --host 0.0.0.0 --port 5000 --gunicorn-opts '--forwarded-allow-ips=*' &")

time.sleep(3) # Tunggu server mlflow naik

# 4. Sambungkan kembali Ngrok ke port 5000
try:
    public_url = ngrok.connect(5000, proto="http")
    print("--- RE-SETUP SERVER MLFLOW BERHASIL ---")
    print("Silakan klik link baru di bawah ini:")
    print(public_url.public_url)
except Exception as e:
    print("Terjadi kesalahan:", e)

--- RE-SETUP SERVER MLFLOW BERHASIL ---
Silakan klik link baru di bawah ini:
https://preppy-qualm-flashcard.ngrok-free.dev


In [ ]:
# Jalankan server MLflow di background port 5000
get_ipython().system_raw("mlflow ui --port 5000 &")

# Gunakan localtunnel untuk mengekspos port 5000 ke internet agar bisa dibuka
!npx localtunnel --port 5000

⠙⠹⠸⠼⠴⠦Need to install the following packages:
localtunnel@2.0.2
Ok to proceed? (y) ^C


### Advanced Feature engineering
1. Temporal Features (Cyclic Transformation): Bulan (month) dan hari (day_of_week) adalah data yang berputar (siklik). Kalau kita pakai angka 1-12 biasa, model menganggap bulan Desember (12) sangat jauh dari Januari (1). Dengan mengubahnya ke Sinus dan Cosinus, model akan tahu bahwa Desember dan Januari itu berdekatan secara waktu.

2. Spatial Feature Proxy: Karena kita tidak punya data koordinat, kita buat proxy tingkat kesejahteraan wilayah berdasarkan kombinasi indikator ekonomi makro yang ada di dataset `(emp.var.rate, cons.price.idx)`


In [9]:
def advanced_feature_engineering(data):
    # Duplikat dataframe agar tidak merusak data asli
    df_feat = data.copy()

    # --- 1. TEMPORAL FEATURES (CYCLIC TRANSFORMATION) ---
    # Mapping bulan ke angka
    month_map = {'mar':3, 'apr':4, 'may':5, 'jun':6, 'jul':7, 'aug':8, 'sep':9, 'oct':10, 'nov':11, 'dec':12}
    df_feat['month_num'] = df_feat['month'].map(month_map)

    # Mapping hari ke angka
    day_map = {'mon':1, 'tue':2, 'wed':3, 'thu':4, 'fri':5}
    df_feat['day_num'] = df_feat['day_of_week'].map(day_map)

    # Transformasi Sine & Cosine untuk Bulan (Total 12 bulan dalam sekitaran penuh)
    df_feat['month_sin'] = np.sin(2 * np.pi * df_feat['month_num'] / 12)
    df_feat['month_cos'] = np.cos(2 * np.pi * df_feat['month_num'] / 12)

    # Transformasi Sine & Cosine untuk Hari (Total 5 hari kerja di dataset)
    df_feat['day_sin'] = np.sin(2 * np.pi * df_feat['day_num'] / 5)
    df_feat['day_cos'] = np.cos(2 * np.pi * df_feat['day_num'] / 5)

    # --- 2. SPATIAL FEATURE PROXY (ECONOMIC CONTEXT) ---
    # Membuat proxy indeks tekanan ekonomi wilayah nasabah tinggal
    # menggabungkan beberapa indikator linear menjadi satu fitur interaksi yang bermakna
    df_feat['economic_stress_index'] = df_feat['emp.var.rate'] * df_feat['cons.price.idx']

    # --- 3. CLEAN UP & DROPPING ORIGINAL COLUMNS ---
    # Hapus kolom asal yang sudah ditransformasikan agar tidak redundan
    df_feat = df_feat.drop(['month', 'month_num', 'day_of_week', 'day_num'], axis=1)

    return df_feat

# menjalankan fungsi engineering
df_engineered = advanced_feature_engineering(df)
print("Bentuk data setelah Advanced Feature Engineering:", df_engineered.shape)
df_engineered[['month_sin', 'month_cos', 'day_sin', 'day_cos', 'economic_stress_index']].head()

Bentuk data setelah Advanced Feature Engineering: (41188, 24)


,month_sin,month_cos,day_sin,day_cos,economic_stress_index
0,0.5,-0.866025,0.951057,0.309017,103.3934
1,0.5,-0.866025,0.951057,0.309017,103.3934
2,0.5,-0.866025,0.951057,0.309017,103.3934
3,0.5,-0.866025,0.951057,0.309017,103.3934
4,0.5,-0.866025,0.951057,0.309017,103.3934


### membangun base line model & Logging Otomatis ke MLflow

- memisahkan data menjadi Train & Test, melakukan encoding pada variabel kategori tersisa, lalu melatih Baseline Model menggunakan LogisticRegression.

- Karena sudah mengaktifkan `mlflow.sklearn.autolog()`. Jadi, ketika melakukan `.fit()`, MLflow akan otomatis merekam semua metrik evaluasi model kita tanpa perlu kita ketik satu-persatu!

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score, log_loss

# 1. Encoding Variabel Kategorikal Tersisa & Target
df_model = df_engineered.copy()

# Encode Target 'y' (no -> 0, yes -> 1)
df_model['y'] = df_model['y'].map({'no': 0, 'yes': 1})

# Encode fitur kategori lainnya dengan One-Hot Encoding
categorical_cols = df_model.select_dtypes(include=['object']).columns
df_model = pd.get_dummies(df_model, columns=categorical_cols, drop_first=True)

# 2. Split Data (Features & Target)
X = df_model.drop('y', axis=1)
y = df_model['y']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# 3. Training Baseline Model di dalam MLflow Run
with mlflow.start_run(run_name="Baseline_Logistic_Regression"):

    # Inisialisasi dan latih model
    # Set max_iter lebih tinggi agar model konvergen
    baseline_model = LogisticRegression(max_iter=1000, random_state=42)
    baseline_model.fit(X_train, y_train)

    # Prediksi untuk evaluasi tambahan jika diperlukan
    y_pred = baseline_model.predict(X_test)
    y_prob = baseline_model.predict_proba(X_test)[:, 1]

    # menhitung metrik advanced untuk performa imbalanced data
    logloss_val = log_loss(y_test, y_prob)
    roc_auc_val = roc_auc_score(y_test, y_prob)

    # Manual log untuk metrik spesifik yang sangat krusial bagi bisnis bank
    mlflow.log_metric("custom_log_loss", logloss_val)
    mlflow.log_metric("custom_roc_auc", roc_auc_val)

    print("--- BASELINE MODEL EVALUATION ---")
    print(classification_report(y_test, y_pred))
    print(f"ROC-AUC Score: {roc_auc_val:.4f}")

2026/06/04 01:24:28 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:440: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERAT

--- BASELINE MODEL EVALUATION ---
              precision    recall  f1-score   support

           0       0.93      0.98      0.95      7310
           1       0.70      0.40      0.51       928

    accuracy                           0.91      8238
   macro avg       0.81      0.69      0.73      8238
weighted avg       0.90      0.91      0.90      8238

ROC-AUC Score: 0.9349


- Nilai Recall-nya cuma 0.40. Artinya, dari total 928 orang yang aslinya mau menaruh deposito, model hanya berhasil menebak 40% di antaranya. Sisanya (60%) lolos tidak terdeteksi. Memiliki potensi kerugian bisnis (opportunity loss)

- Log mendeteksi `lbfgs failed to converge`. Ini karena data kamu belum di-scaling (standarisasi) sedangkan model Logistic Regression sangat sensitif terhadap perbedaan skala angka

### Tree-Based Modeling (LightGBM) & Autologging

In [11]:
import lightgbm as lgb
from sklearn.metrics import classification_report, roc_auc_score, precision_recall_curve, auc
import mlflow.lightgbm

# mengaktifkan autolog khusus untuk LightGBM
mlflow.lightgbm.autolog()

with mlflow.start_run(run_name="Advanced_LightGBM_Model"):

    # Inisialisasi model LightGBM dengan penanganan class imbalance (is_unbalance=True)
    lgb_model = lgb.LGBMClassifier(
        n_estimators=100,
        learning_rate=0.05,
        random_state=42,
        is_unbalance=True, # Fitur krusial untuk data jomplang/imbalanced
        verbosity=-1
    )

    # Training model
    lgb_model.fit(X_train, y_train)

    # Prediksi
    y_pred_lgb = lgb_model.predict(X_test)
    y_prob_lgb = lgb_model.predict_proba(X_test)[:, 1]

    # menghitung PR-AUC (Precision-Recall AUC) -> Metrik emas untuk data imbalanced
    precision, recall, _ = precision_recall_curve(y_test, y_prob_lgb)
    pr_auc_val = auc(recall, precision)
    roc_auc_lgb = roc_auc_score(y_test, y_prob_lgb)

    # Log metrik tambahan ke MLflow secara manual
    mlflow.log_metric("PR_AUC", pr_auc_val)
    mlflow.log_metric("ROC_AUC", roc_auc_lgb)

    print("--- LIGHTGBM MODEL EVALUATION ---")
    print(classification_report(y_test, y_pred_lgb))
    print(f"ROC-AUC Score: {roc_auc_lgb:.4f}")
    print(f"PR-AUC Score: {pr_auc_val:.4f}")

2026/06/04 01:25:42 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:440: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/06/04 01:25:44 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/lib/python3.12/dist-packages/mlflow/types/utils

--- LIGHTGBM MODEL EVALUATION ---
              precision    recall  f1-score   support

           0       0.99      0.86      0.92      7310
           1       0.46      0.94      0.62       928

    accuracy                           0.87      8238
   macro avg       0.73      0.90      0.77      8238
weighted avg       0.93      0.87      0.89      8238

ROC-AUC Score: 0.9557
PR-AUC Score: 0.7042


model berhasil mendeteksi 94% nasabah yang berpotensi mengambil deposito (dibandingkan model baseline yang cuma 40%).

### Pseudo-Labeling (Semi-Supervised)

Data nasabah baru sangat banyak tapi belum tahu mereka mau langganan atau tidak (Unlabeled Data).

maka perlu mengambil sebagian kecil dari X_test seolah-olah sebagai data tanpa label, lalu meminta model LightGBM untuk melabelinya secara otomatis (Pseudo-Label), dan menggabungkannya kembali ke data training untuk meningkatkan performa model

In [12]:
with mlflow.start_run(run_name="Pseudo_Labeling_Experiment"):

    # 1. Simulasikan "Unlabeled Data" dari sebagian X_test
    X_unlabeled = X_test.sample(frac=0.5, random_state=42)

    # 2. Gunakan model LightGBM yang sudah terlatih untuk memprediksi probabilitas data unlabeled
    pseudo_probs = lgb_model.predict_proba(X_unlabeled)[:, 1]

    # 3. Tentukan Threshold Tinggi (Hanya ambil prediksi yang model SANGAT YAKIN > 90% atau < 10%)
    high_confidence_yes = pseudo_probs > 0.90
    high_confidence_no = pseudo_probs < 0.10

    # menggabungkan data yang meyakinkan
    pseudo_features = X_unlabeled[high_confidence_yes | high_confidence_no]
    pseudo_labels = np.where(pseudo_probs[high_confidence_yes | high_confidence_no] > 0.5, 1, 0)

    print(f"Jumlah pseudo-label berkualitas tinggi yang didapat: {len(pseudo_features)} sampel")

    # 4. Augmentasi / Gamenggabungkan ke Data Training Asli
    X_train_augmented = pd.concat([X_train, pseudo_features], axis=0)
    y_train_augmented = np.concatenate([y_train, pseudo_labels])

    # 5. melatih Ulang Model LightGBM dengan Data Baru yang Lebih Kaya
    final_model = lgb.LGBMClassifier(n_estimators=100, learning_rate=0.05, random_state=42, is_unbalance=True, verbosity=-1)
    final_model.fit(X_train_augmented, y_train_augmented)

    # 6. mengevaluasi Akhir pada Sisa Data Test yang Bersih
    y_prob_final = final_model.predict_proba(X_test)[:, 1]
    roc_auc_final = roc_auc_score(y_test, y_prob_final)

    mlflow.log_metric("Final_Augmented_ROC_AUC", roc_auc_final)
    mlflow.log_param("pseudo_labels_added", len(pseudo_features))

    print("\n--- FINAL MODEL (WITH PSEUDO-LABELING) EVALUATION ---")
    print(f"Final ROC-AUC Score: {roc_auc_final:.4f}")

2026/06/04 01:25:55 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:440: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."


Jumlah pseudo-label berkualitas tinggi yang didapat: 2950 sampel


2026/06/04 01:25:57 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:440: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/06/04 01:25:59 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/lib/python3.12/dist-packages/mlflow/types/utils


--- FINAL MODEL (WITH PSEUDO-LABELING) EVALUATION ---
Final ROC-AUC Score: 0.9554


Pseudo-Labeling berhasil memanen 2950 sampel data berkualitas tinggi untuk memperkuat kecerdasan model

### save the best model

In [13]:
# 2. menyimpan di dalam folder di Gdrive
base_path = "/content/drive/MyDrive/ML_Project/MLops_Bank_Project"
os.makedirs(base_path, exist_ok=True)

print(f"Google Drive berhasil terhubung dengan baik! File akan disimpan di: {base_path}")

Google Drive berhasil terhubung dengan baik! File akan disimpan di: /content/drive/MyDrive/ML_Project/MLops_Bank_Project


In [14]:
import joblib
import json

# Definisikan path lengkap ke Google Drive
model_path = os.path.join(base_path, 'bank_deposit_model_v1.pkl')
columns_path = os.path.join(base_path, 'model_columns.json')

# 1. Simpan objek model ke Google Drive
joblib.dump(final_model, model_path)

# 2. Simpan skema kolom ke Google Drive
with open(columns_path, 'w') as f:
    json.dump(list(X.columns), f)

print("✅ Model dan Skema Kolom AMAN & PERMANEN tersimpan di Google Drive !")

✅ Model dan Skema Kolom AMAN & PERMANEN tersimpan di Google Drive !


### save model otomasi di hugging face

In [15]:
!pip install huggingface_hub -q

In [17]:
from huggingface_hub import HfApi, login
import json
import joblib

# 1. Token HF
HF_TOKEN = "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"
login(token=HF_TOKEN)

# 2. repo name
repo_id = "nikenlarash22/bank-model-lightbgm"

api = HfApi()
# Membuat repository model baru di Hugging Face jika belum ada
try:
    api.create_repo(repo_id=repo_id, repo_type="model", private=False)
    print(f"Repository {repo_id} berhasil dibuat!")
except Exception as e:
    print("Repository sudah ada atau terjadi error:", e)

# 3. Upload File Model dan Skema Kolom ke Hugging Face
# mengambil dari folder Google Drive
api.upload_file(
    path_or_fileobj="/content/drive/MyDrive/ML_Project/MLops_Bank_Project/bank_deposit_model_v1.pkl",
    path_in_repo="bank_deposit_model_v1.pkl",
    repo_id=repo_id,
    repo_type="model"
)

api.upload_file(
    path_or_fileobj="/content/drive/MyDrive/ML_Project/MLops_Bank_Project/model_columns.json",
    path_in_repo="model_columns.json",
    repo_id=repo_id,
    repo_type="model"
)

print("✅ Sukses! Model dan Skema Kolom sudah masuk di Hugging Face Hub!")

Repository sudah ada atau terjadi error: Client error '409 Conflict' for url 'https://huggingface.co/api/repos/create' (Request ID: Root=1-6a20e09b-68c6c21528926af001a51987;169b3ca0-7f41-405c-a433-be2554930f28)
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409

You already created this model repo: nikenlarash22/bank-model-lightbgm


Processing Files (0 / 0)      : |          |  0.00B /  0.00B            

New Data Upload               : |          |  0.00B /  0.00B            

  ...bank_deposit_model_v1.pkl:  81%|########1 |  285kB /  351kB            

✅ Sukses! Model dan Skema Kolom sudah masuk di Hugging Face Hub!
